### PR TITLE
PP-5042 Make cancelling use smaller, tighter transactions

### PR DESCRIPTION
**Pull request by alexbishop1**
When cancelling a charge, don’t do the entire operation, including sending a cancellation message to the gateway, inside a single transaction because this gives lots of time for something else to come in and change the charge in the mean time. Instead, have transactions tightly around just the database updates and do the communication with the gateway outside a transaction. This means we will atomically change the status to a `READY` state, then communicate with the gateway, then atomically change the status to a terminal cancellation state. If anything else tries to change the charge while we’re talking to the gateway, it will observe the charge in a `READY` state and know to not try to do that.

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/charge/service/ChargeCancelService.java [hunk 1]
```diff
@@ -127,20 +124,20 @@ private void nonGatewayCancel(ChargeEntity chargeEntity, StatusFlow statusFlow)
...
-        chargeEventDao.persistChargeEventOf(chargeEntity);
+
+        setChargeStatusTo(chargeEntity, completeStatus);
     }
 
     private boolean gatewayIsNotAwareOfCharge(ChargeEntity chargeEntity) {
         return nonGatewayStatuses.contains(ChargeStatus.fromString(chargeEntity.getStatus()));
     }
 
+    @Transactional
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/charge/service/ChargeCancelService.java]: This method should be public for \@Transactional annotation to work as per https://github.com/google/guice/wiki/Transactions

File: src/main/java/uk/gov/pay/connector/charge/service/ChargeCancelService.java [hunk 2]
```diff
@@ -154,6 +151,12 @@ private void prepareForTerminate(ChargeEntity chargeEntity, StatusFlow statusFlo
                 chargeEntity.getGatewayAccount().getType(),
                 newStatus);
 
+        setChargeStatusTo(chargeEntity, newStatus);
+    }
+
+    @Transactional
+    private void setChargeStatusTo(ChargeEntity chargeEntity, ChargeStatus chargeStatus) {
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/charge/service/ChargeCancelService.java]: Same as above for public method

**Review by kbottla — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/pay/connector/charge/service/ChargeCancelService.java (hunk 1 above)
Comment (alexbishop1) [reply src/main/java/uk/gov/pay/connector/charge/service/ChargeCancelService.java]: Oops! Will fix

**Review by kbottla — DISMISSED** [review]
LGTM 👍

**Comment by alexbishop1** [comment]
End-to-end tests alerted us that this change wasn’t persisting the statuses correctly. Fixed that by making the code retrieve the `ChargeEntity` from the DAO between transactions (which required unit test changes).

**Review by kbottla — APPROVED** [review]
LGTM!

**Pull request merged**